include/unifex/bind_back.hpp

```diff
@@ -44,23 +88,57 @@ struct _apply_target<Cpo, ArgN...>::type {
     noexcept(std::is_nothrow_constructible_v<std::tuple<remove_cvref_t<ArgN>...>, CN...>) 
     : argN_((CN&&) cn...) {}
 
-  template <typename Target>
+  template <typename Target, typename... _ArgN>
+  using _result_t =
+    typename meta_quote1_<callable_result_t>::template apply<Cpo, Target, _ArgN...>;
+
+  template (typename Target)
+    (requires (!derived_from<Target, _result_base>) AND callable<Cpo, Target, ArgN const&...>)
+  auto operator()(Target&& target) const &
+    noexcept(
+      is_nothrow_callable_v<Cpo, Target, ArgN const&...>) 
+    -> _result_t<Target, ArgN const&...> {
+    return std::apply(_apply_fn<Cpo, Target>{&target}, argN_);
+  }
+  template (typename Target)
+    (requires (!derived_from<Target, _result_base>) AND callable<Cpo, Target, remove_cvref_t<ArgN>...>)
   auto operator()(Target&& target) &&
     noexcept(
-      is_nothrow_callable_v<Cpo, Target, ArgN...>) 
-    -> callable_result_t<Cpo, Target, ArgN...> {
-    return std::apply([&](auto&&... argN){
-      return Cpo{}((Target&&) target, (ArgN&&) argN...);
-    }, std::move(argN_));
+      is_nothrow_callable_v<Cpo, Target, remove_cvref_t<ArgN>...>) 
+    -> _result_t<Target, remove_cvref_t<ArgN>...> {
+    return std::apply(_apply_fn<Cpo, Target>{&target}, std::move(argN_));
   }
 
-  template <typename Target>
+  template (typename Target)
+    (requires (!derived_from<Target, _result_base>) AND callable<Cpo, Target, remove_cvref_t<ArgN>...>)
   friend auto operator|(Target&& target, type&& self) 
     noexcept(
-      is_nothrow_callable_v<Cpo, Target, ArgN...>) 
-    -> callable_result_t<Cpo, Target, ArgN...> {
+      is_nothrow_callable_v<Cpo, Target, remove_cvref_t<ArgN>...>) 
+    -> _result_t<Target, remove_cvref_t<ArgN>...> {
     return std::move(self)((Target&&) target);
   }
+  template (typename Target)
+    (requires (!derived_from<Target, _result_base>) AND callable<Cpo, Target, ArgN const&...>)
+  friend auto operator|(Target&& target, const type& self) 
+    noexcept(
+      is_nothrow_callable_v<Cpo, Target, ArgN const&...>) 
+    -> _result_t<Target, ArgN const&...> {
+    return self((Target&&) target);
+  }
+
+  template (typename Other, typename Self)
+    (requires derived_from<Other, _result_base> AND same_as<remove_cvref_t<Self>, type>)
+  friend _result<detail::compose, Other, Self> 
+  operator|(Other&& other, Self&& self) 
+    noexcept(
+      std::is_nothrow_constructible_v<
+        _result<detail::compose, Other, Self>,
+        Other,
+        Self
+      >) {
+    return _result<detail::compose, Other, Self>{
+      (Other&&)other, (Self&&)self};
+  }
```

Comment:
I think overload resolution is still going to cause the compiler to compute a bunch of different return types here unnecessarily. This needs some thought. I would say, fix the clang 10 issue and the `derived_from<Other, _result_base>` constraint above and then this is good to go. I can try to re-express it to compute the fewest number of types later.